Fix bug with --clean flag

jwakre opened the pull request:
Issue #564 identified an issue with the --clean flag.  

* This change **impacts format bumps**, so we need to be careful that nothing breaks. This PR has been updated so that LAST_VER overrides PREVIOUS_MIX_VERSION, so format bumps should behave similarly.

jwakre commented on bat/lib/mixerlib.bash:
```diff
@@ -30,10 +30,10 @@ mixer-init-stripped-down() {
 }
 mixer-mixversion-update() {
-  mixer $MIXARGS versions update --mix-version $1 --native=true
+  sudo -E mixer $MIXARGS versions update --mix-version $1 --native=true
```
The added dependence on sudo for this command could cause users problems.

rchiossi commented:
I quite dislike this approach because it changes the meaning on LAST_VER. This was supposed to be the last good bundles build, not the previous. The problem is that if you are rebuilding over a build that succeeded we currently have no track of what the previous one was. On top of that, the added sudo dependency just further emphasize that this is not the best way to track that information.

I believe a cleaner approach would be to use `mixer.state` to track previous version number. We would them update that value when `mixer versions update` is called and use that information with `--clean` instead of LAST_VER. This way we don't add the extra sudo dependency and minimize the impact on code that relies on the current LAST_VER.

Note that the contents of `update/image` are not part of the published content, so we can change it if we see fit. If we could move to use mixer.state only and ditch LAST_VER in the future, it would be even better.

jwakre commented:
@rchiossi To rework this PR so the previous version is tracked by mixer.state, I plan on making the following changes:

1. Create a ```PREVIOUS``` section inside mixer.state to track the previous version and **increment the minor mixer.state version**. The ```PREVIOUS```  version will be updated by the ```mixer versions update``` command, increment flag, and format bumps.

2. Add a config/state_convert.go file to update the mixer.state file for version changes similarly to config/config_convert.go. This will be useful to insert the ```PREVIOUS``` field into the mixer.state file for users that upgrade Mixer and have an existing mixer.state file.

3. When unset, the  ```PREVIOUS```  field will default to the ```LAST_VER``` and fallback to 0 when  ```LAST_VER``` does not exist.

4. The value of ```PREVIOUS``` will be used to set the previous version for manifests instead of ```LAST_VER```.

5. Format bumps will override the ```PREVIOUS``` field with ```LAST_VER``` so that format bump behavior remains unchanged.

The values of ```PREVIOUS``` and ```LAST_VER``` will be identical except for the following case:
* ```mixer build update``` or ```mixer build all``` was ran without the increment flag
  * Since the ```mixversion``` has not been updated, the value of ```PREVIOUS``` cannot be updated. This resolves the bug with the clean flag.

rchiossi commented:
> 1. Create a `PREVIOUS` section inside mixer.state to track the previous version and **increment the minor mixer.state version**. The `PREVIOUS`  version will be updated by the `mixer versions update` command, increment flag, and format bumps.

Use a more descriptive name here since it is likely that we will have versions for multiple different things here in the future (e.g. upstreamver and mixver), so it is good to make it explicit that this refers to the mix version.

> The values of `PREVIOUS` and `LAST_VER` will be identical except for the following case:
> 
> * `mixer build update` or `mixer build all` was ran without the increment flag
>   
>   * Since the `mixversion` has not been updated, the value of `PREVIOUS` cannot be updated. This resolves the bug with the clean flag.

They will also be different after `build bundles` and before `build update`. This might be relevant for format bumps, keep this in mind when checking that code.

Other than that, the plan looks good!

jwakre commented:
I updated the PR with my previously outlined changes. I will continue testing them to confirm everything works as expected.

jwakre commented:
I've performed additional testing on upstream/downstream format bumps, config conversions, builds, and clean builds. The PREVIOUS_MIX_VERSION field in mixer.state has been adjusted as expected and update content has been generated correctly.

jwakre commented:
After testing Mixin, the state conversion saves the updated state file in the wrong directory. I will update this PR to address this issue with Mixin.

jwakre commented on config/state.go:
```diff
@@ -74,6 +77,16 @@ func (state *MixState) loadDefaultFormat() {
 
+func (state *MixState) loadDefaultPreviousMixVer() {
+	/* The LAST_VER is the default for PREVIOUS_MIX_VERSION */
+	lastVer, err := ioutil.ReadFile("update/image/LAST_VER")
```
Typically, mixes are built in the current working directory, so relative paths will usually work. Since Mixin uses an alternative mix path (/usr/share/mix), a path relative to the current directory will likely be incorrect.

Also, there are existing cases on the master branch where state files use relative paths which will cause issues for Mixin.

jwakre commented:
I updated the PR to make the following for changes to Mixin:

* The value of PREVIOUS_MIX_VERSION is overridden by LAST_VER for Mixin. This keeps Mixin behavior consistent. 
* The working directory for Mixin is changed to the mix directory (/usr/share/mix) before interacting with the mixer.state file. This solves an issue where the mixer.state file was saved to the current working directory instead of the mix directory.

rchiossi commented on config/state_convert.go:
```diff
@@ -0,0 +1,73 @@
+
+func (state *MixState) parseVersionAndConvert() error {
+	// Reset version for files without versioning
+	state.version = "0.0"
```
This is not needed for mixer.state because unlike builder.conf, it's file always had the `#VERSION` header, so there is no valid mixer.state file without versioning.

rchiossi commented on config/state_convert.go:
```diff
@@ -0,0 +1,73 @@
+
+	if err := helpers.CopyFile(state.filename+".bkp", state.filename); err != nil {
+		return err
+	}
```
This is done for builder.conf because it is a file that holds user config. Since mixer.state should only be modified by mixer commands and is considered and "internal storage", there is no need to do a backup.

rchiossi commented on bat/tests/state-conversion/run.bats:
```diff
@@ -0,0 +1,57 @@
+  do
+    copy_test_state $f
+
+    mixer versions update --mix-version 20
```
For simple conversion, it would be better to use a command that does not change mixer.state, like `mixer versions --offline` for instance. Since `mixer versions update` will overwrite that value, it doesn't guarantee that the value is properly converted. It doesn't make any difference if you convert from v1.0 to v1.1 since the value is not there, but this wont be true for future versions.

rchiossi commented on bat/tests/clean-rebuild/run.bats:
```diff
@@ -0,0 +1,31 @@
+
+  # Check that PREVIOUS_MIX_VERSION in mixer.state is correct
+  test $(sed -n 's/[ ]*PREVIOUS_MIX_VERSION[ ="]*\([0-9]\+\)[ "]*/\1/p' mixer.state) -eq 10
+}
```
Could we reproduce the bug with the missing entry as part of this test as well?

jwakre replied:
sure!

jwakre commented:
> I'm still a bit reluctant on the way we rely on `LAST_VER` to set the value for `PREVIOUS_MIX_VER`. The correct value to use would be the version of the last successful `mixer build update` rather than the last successful `mixer build bundles`, but that is a value that we don't keep track.

`LAST_VER` is initialized to 0 by `mixer build bundles`, but otherwise it is updated by `mixer build update` (and format bumps). Although, `LAST_VER` will incorrectly initialize `PREVIOUS_MIX_VER` when `LAST_VER` has been updated, but mixversion has not ( `mixer build update` w/o the increment flag).

> This information is somewhat available in the MoM of the most recent version, but but there is no way to tell if the highest version number in the www folder is actually a good build.
> 
> For now, the patch is good as it is. I'd like to have a follow-up on it though, where we improve our overall version tracking. We should sit down and come up with a plan for it.

Sounds good!

jwakre commented on config/state_convert.go:
```diff
@@ -0,0 +1,73 @@
+	return state.convertCurrent()
+}
+
+func (state *MixState) convertCurrent() error {
```
Ok, I think consistency makes it easier to understand both configs, so I will keep this the same.

jwakre commented:
Thanks for the feedback. I made the following changes:

*  Removed state file version reset (setting to 0.0)
*  Removed backup statefile
*  Replaced `mixer versions update` with `mixer versions --offline`
* Added a negative test case for the clean build bat test
